qbft/round_change.go

```diff
@@ -0,0 +1,347 @@
+			signedRoundChange.Message.Round,
+		) == nil {
+			// not returning error, no need to
+			return msg, nil
+		}
+	}
+	return nil, nil
+}
+
+// isReceivedProposalJustificationForLeadingRound - returns nil if we have a quorum of round change msgs and highest justified value for leading round
+func isReceivedProposalJustificationForLeadingRound(
+	state *State,
+	config IConfig,
+	roundChangeMsg *SignedMessage,
+	roundChanges []*SignedMessage,
+	valCheck ProposedValueCheckF,
+	newRound Round,
+) error {
+	rcData, err := roundChangeMsg.Message.GetRoundChangeData()
+	if err != nil {
+		return errors.Wrap(err, "could not get round change data")
+	}
+
+	if err := isReceivedProposalJustification(
+		state,
+		config,
+		roundChanges,
+		rcData.RoundChangeJustification,
+		roundChangeMsg.Message.Round,
+		rcData.PreparedValue,
+		valCheck); err != nil {
+		return err
+	}
+
+	if proposer(state, config, roundChangeMsg.Message.Round) != state.Share.OperatorID {
+		return errors.New("not proposer")
+	}
+
+	currentRoundProposal := state.ProposalAcceptedForCurrentRound == nil && state.Round == newRound
+	futureRoundProposal := newRound > state.Round
+
+	if !currentRoundProposal && !futureRoundProposal {
+		return errors.New("proposal round mismatch")
+	}
+
+	return nil
+}
+
+// isReceivedProposalJustification - returns nil if we have a quorum of round change msgs and highest justified value
+func isReceivedProposalJustification(
+	state *State,
+	config IConfig,
+	roundChanges, prepares []*SignedMessage,
+	newRound Round,
+	value []byte,
+	valCheck ProposedValueCheckF,
+) error {
+	if err := isProposalJustification(
+		state,
+		config,
+		roundChanges,
+		prepares,
+		state.Height,
+		newRound,
+		value,
+		valCheck,
+	); err != nil {
+		return errors.Wrap(err, "proposal not justified")
+	}
+	return nil
+}
+
+func validRoundChange(state *State, config IConfig, signedMsg *SignedMessage, height Height, round Round) error {
+	if signedMsg.Message.MsgType != RoundChangeMsgType {
+		return errors.New("round change msg type is wrong")
+	}
+	if signedMsg.Message.Height != height {
+		return errors.New("round change Height is wrong")
+	}
+	if signedMsg.Message.Round != round {
+		return errors.New("msg round wrong")
+	}
+	if len(signedMsg.GetSigners()) != 1 {
+		return errors.New("round change msg allows 1 signer")
+	}
+
+	if err := signedMsg.Signature.VerifyByOperators(signedMsg, config.GetSignatureDomainType(), types.QBFTSignatureType, state.Share.Committee); err != nil {
+		return errors.Wrap(err, "round change msg signature invalid")
+	}
+
+	rcData, err := signedMsg.Message.GetRoundChangeData()
+	if err != nil {
+		return errors.Wrap(err, "could not get roundChange data ")
+	}
+	if err := rcData.Validate(); err != nil {
+		return errors.Wrap(err, "roundChangeData invalid")
+	}
+
+	// Addition to formal spec
+	// We add this extra tests on the msg itself to filter round change msgs with invalid justifications, before they are inserted into msg containers
```

Comment:
Does this imply that each ROUND-CHANGE message piggybacked in the proposal also piggybacks PREPARE messages? If so, then this makes total communication complexity of the protocol O(N^3).

Reply:
Why is that? message complexity is how many times that message bounces in the network right? so Node 1 needs to broadcast it to all other nodes and so on